src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,487 @@
+    """Surrogate for the :func:`trust_region` enriching with both the primal and dual solutions.
+
+    Parameters
+    ----------
+    reductor
+        The reductor used to generate the reduced order models and estimate the output error.
+    initial_guess
+        The |parameter values| containing an initial guess for the optimal parameter value.
+    """
+
+    def extend(self, mu):
+        """Extend the current ROM for new |parameter values| with primal and dual solutions.
+
+        Parameters
+        ----------
+        mu
+            The `Mu` instance for which an extension is computed.
+        """
+        with self.logger.block('Extending the basis with primal and dual...'):
+            fom = self.reductor.fom
+            U_h_mu = fom.solve(mu)
+            jacobian = fom.output_functional.jacobian(U_h_mu, self.reductor.fom.parameters.parse(mu))
+            dual_solutions = fom.solution_space.empty()
+            for d in range(fom.dim_output):
+                dual_problem = fom.with_(operator=fom.operator.H, rhs=jacobian.H.as_range_array(mu)[d])
+                P_h_mu = dual_problem.solve(mu)
+                dual_solutions.append(P_h_mu)
+
+            self.fom_evaluations += 1 + fom.dim_output
+            self.new_reductor = deepcopy(self.reductor)
+            try:
+                self.new_reductor.extend_basis(U_h_mu)
+            except ExtensionError:
+                self.new_reductor = self.reductor
+            # it can happen that primal is successful but duals are not.
+            try:
+                self.new_reductor.extend_basis(dual_solutions)
+            except ExtensionError:
+                pass
+            self.new_rom = self.new_reductor.reduce()
+
+class QuadraticOutputTRSurrogate(BasicTRSurrogate):
+    """Surrogate for :func:`trust_region` with only the primal enrichment naive output estimate.
+
+    NOTE: This specialized TRSurrogate should soon be made obsolete when the quadratic
+          output estimation is included into a reductor.
+
+    Parameters
+    ----------
+    reductor
+        The reductor used to generate the reduced order models and estimate the output error.
+    initial_guess
+        The |parameter values| containing an initial guess for the optimal parameter value.
+    continuity_estimator_output
+        Estimation for the continuity constant of the quadratic output functional.
+    """
+
+    def __init__(self, reductor, initial_guess, continuity_estimator_output):
+        self.continuity_estimator_output = continuity_estimator_output
+        super().__init__(reductor, initial_guess)
+
+    def estimate_output_error(self, mu):
+        self.rom_evaluations += 1
+        U, pr_err = self.rom.solve(mu, return_error_estimate=True)
+        cont_est = self.continuity_estimator_output
+        cont = cont_est.evaluate(self.parameters.parse(mu)) if hasattr(cont_est, 'evaluate') else cont_est
+        return cont * (pr_err * (2 * self.rom.l2_norm(U) + pr_err))
```

Comment:
This assumes that fom/rom actually have an `l2` inner product. I think it would be better to specify the product used to compute the norm in `__init__`. Also it should be documented w.t.r. which norm the continuity constant is estimated.